### DISCUSSION TITLE
How can i use reactive value in `initialValues ` in `useDragAndDrop` in vuejs

### DISCUSSION THREAD

**Discussion by alokVishu** [formkit/drag-and-drop#52] [General]
Hello there,

I am trying to create a Kanban app using a drag-and-drop plugin. I have to force update the component when I modify the initialValues array. 

Is it possible to use a ref (reactive) value instead of a plain array?

 ```vue
const initialValues = [...] // want to use ref 
 const [kanbanWrapper, todo] = useDragAndDrop(initialValues, { plugins: [animations()] })
 ```

**Comment by sashamilenkovic** [comment]
@alokVishu Hey there! You can use the alternative hook, `dragAndDrop`: https://drag-and-drop.formkit.com/#draganddrop

**Reply by alokVishu** [reply]
@sashamilenkovic, Thank you for the reply.

I couldn't find any usage example of the `dragAndDrop` hook.

**Reply by sashamilenkovic** [reply]
@alokVishu Yeah, it would probably be worth adding an example for it. In the meanwhile, here's a quick example:

```ts
<script setup lang="ts">
import { ref } from "vue";
import { dragAndDrop } from "@formkit/drag-and-drop/vue";

const items = ref(["Item 1", "Item 2", "Item 3"]);

const parent = ref();

dragAndDrop({
  parent,
  values: items,
});
</script>

<template>
  <div ref="parent">
    <div v-for="item in items" :key="item" class="item">
      {{ item }}
    </div>
  </div>
</template>
```

**Reply by alokVishu** [reply]
Thank you for your assistance. I will give it a try and let you know the results.

**Reply by alokVishu** [reply]
It works for me. Thanks.

**Reply by alokVishu** [reply]
Hello @sashamilenkovic,

I am currently experiencing a problem with the dragAndDrop hook. I am trying to create a Kanban board and I need to use data from a database. However, when I dynamically manipulate the data and then re-fetch it, the dragAndDrop hook does not get re-initialized.

Is there a way to re-initialize the dragAndDrop hook when values are updated?

```

//fetching data from database 
const {data} = useFetch('...')

// clone the fetch data in local ref variable
const localKanbanData = ref(JSON.parse(JSON.stringify(data)))

// if data updated it re-clone it
watch(data, () =>{
localKanbanData.value = JSON.parse(JSON.stringify(data))
})

// 👉 initialize the drag and drop
dragAndDrop({
  parent: kanbanWrapper,
  values: localKanbanData, // re-initialize when this reactive data update 
  dragHandle: '.drag-handler',
})
```

A video has been attached for reference. Can you help me resolve the issue?

Thanks

https://github.com/formkit/drag-and-drop/assets/80033712/0b69be95-6564-4941-9b54-eb61d88f839c

**Reply by sashamilenkovic** [reply]
Whoops, put my response in the wrong place: @alokVishu So just so I have this clear, are you recreating a new ref every time you fetch data? As long as the reactive ref is bound to the list of items you are iterating over, you shouldn't need to reinitialize `dragAndDrop` (unless you are trying to update part of the config). I'd be happy to look into this further if you could provide some sort of reproduction.

**Reply by alokVishu** [reply]
Hello @sashamilenkovic,

Thank you for your response. I have created a minimal reproduction that manipulates the data using timeout and when you drag the updated drag item it gets duplicated. For more info refer video and code:


https://github.com/formkit/drag-and-drop/assets/80033712/9e49bee2-ae17-4397-9175-df40d1a89b40

[drag-drop-issue.zip](https://github.com/formkit/drag-and-drop/files/14861777/drag-drop-issue.zip)

**Reply by sashamilenkovic** [reply]
@alokVishu Perfect reproduction! Thank you. Yes, I see the issue. The data is being updated, but not in a way that triggers a rerender (because what is getting updated is the title whereas the key of the list items is the id). In your case, you could absolutely reinvoke `dragAndDrop` (it is idempotent) after updating your ref. That should resolve your issue.

**Reply by alokVishu** [reply]
hey @sashamilenkovic, have you provided anything that reinvokes the `dragAndDrop` instead of doing it like below?

```vue
const reInvokeDraggable = (() => {
  dragAndDrop({
    parent: refKanbanBoard,
    values: localIds,
    group: props.groupName,
    plugins: [animations()],
    draggable: (child: { classList: { contains: (arg0: string) => any } }) => child.classList.contains('kanban-card'),
  })
})();

watch(localIds,reInvokeDraggable)
```
Thanks

**Reply by sashamilenkovic** [reply]
@alokVishu I think in your case, you could just call the remap function from core.

```js
import { remapNodes } from "@formkit/drag-and-drop";

...

// Pass the parent element itself, not the ref
remapNodes(kanbanWrapper.value);
```

**Reply by alokVishu** [reply]
Great, @sashamilenkovic! Let's give it a try.

**Reply by alokVishu** [reply]
@sashamilenkovic, thank you so much it works for me.

**Reply by alokVishu** [reply]
Hello @sashamilenkovic,

Thanks, I have created a Kanban app using this great library, but receive console warnings while moving drag items. The animation sometimes behaves strangely.

```bash
KanbanItems.vue:94 The number of enabled nodes does not match the number of values.
```

https://github.com/formkit/drag-and-drop/assets/80033712/8eb2c6a4-4b1d-4da9-95e7-eddbd71a1982

**Reply by sashamilenkovic** [reply]
@alokVishu Hard to say fro the recording. Typically though, receiving that console warning is a good indication that the values are not aligning with the number of enabled nodes. Anytime you see this warning, it could lead to side effects like with you're seeing with the animations. Would you mind sharing a reproduction that I could take a look at?

**Reply by alokVishu** [reply]
Hello @sashamilenkovic, we appreciate your prompt response and assistance. Below is the minimum reproduction required for the issue.
[drag-drop-animation-issue.zip](https://github.com/formkit/drag-and-drop/files/15057903/drag-drop-animation-issue.zip)

https://github.com/formkit/drag-and-drop/assets/80033712/4e56185e-c63b-42ed-b8e6-8203d4450673

**Reply by sashamilenkovic** [reply]
@alokVishu Okay, appreciate the reproduction. I will take a look at this when I get a chance and get back to you.

**Reply by sashamilenkovic** [reply]
@alokVishu Okay, so on a quick look, I get the error `Maximum recursive updates exceeded`, when I either add an item or sort one. This will definitely cause issues. In your `AboutView.vue`, I removed the logic in `updateItems` and it seems to resolve the recursion issue and the sorting looks like it works now. 

Also, I noticed that when you click "add new item" in `In Review`, it adds the item to the `Done` list and we get the console.warn('The number of enabled nodes does not match the number of values.'). Just wanted to point that out.

**Reply by alokVishu** [reply]
Hello @sashamilenkovic,

I hope you are doing well. I have a question about the `handleEnd` function. This function gets called when a dragged or touchend event occurs. I was wondering if there is a similar function that gets called when we receive a task from another board in shareable boards.

**Reply by sashamilenkovic** [reply]
@alokVishu Hey there! I'm doing well, how about yourself? 

I would suggest maybe overriding the `performTransfer` (similar to how you would the `handleEnd`). That would give you an event listener for anytime an element is actually transferred to another board. Or, you could not override any events, and just create some sort of watcher for your list values to know when your list values have been updated.

**Reply by alokVishu** [reply]
Hello @sashamilenkovic,
Thanks for the reply.
I tried using performTransfer, but it only copies the item instead of transferring it. I have attached a video demonstrating the issue.

https://github.com/formkit/drag-and-drop/assets/80033712/6e078fc6-6876-479c-adc6-882ac552f0dd

**Reply by sashamilenkovic** [reply]
@alokVishu You need to import `performTransfer` and call it like you do for `handleEnd`

**Reply by alokVishu** [reply]
Oh like this? 
```
dragAndDrop({
  parent: refKanbanBoard,
  values: localIds,
  group: props.groupName,
  draggable: child => child.classList.contains('kanban-card'),
  plugins: [animations()],
  performTransfer(state, data) {
    performTransfer(state, data)
  },
})
```

Thanks now it works.

**Reply by sashamilenkovic** [reply]
Right. The idea is that you can overwrite the existing functionality, or, use it as an event listener to add functionality either before or after the core `performTransfer` is invoked.
